File: src/easy/single_number.zig
const std = @import("std");
const expectEqualSlices = std.testing.expectEqualSlices;
const expectEqual = std.testing.expectEqual;

pub fn singleNumber(nums: []const i32) i32 {
    var result: i32 = 0;
    for (nums) |num| {
        result ^= num;
    }
    return result;
}

test "decreasing order lower value non-repeating" {
    const nums = [_]i32{ 2, 2, 1 };
    const expected: i32 = 1;
    const actual = singleNumber(&nums);
    try expectEqual(expected, actual);
}
test "unordered list, higher value non-repeating" {
    const nums = [_]i32{ 4, 1, 2, 1, 2 };
    const expected: i32 = 4;
    const actual = singleNumber(&nums);
    try expectEqual(expected, actual);
}
test "single element array" {
    const nums = [_]i32{1};
    const expected: i32 = 1;
    const actual = singleNumber(&nums);
    try expectEqual(expected, actual);
}

File: src/easy/single_number.rs
pub fn single_number(nums: Vec<i32>) -> i32 {
        nums.into_iter().fold(0, |acc, x| acc ^ x)
}
#[cfg(test)]
mod tests {
    use super::single_number;
    use rstest::rstest;
    #[rstest]
    #[case(vec![2,2,1], 1)]
    #[case(vec![4,1,2,1,2], 4)]
    #[case(vec![1], 1)]
    fn test_single_number(#[case] nums: Vec<i32>, #[case] expected: i32) {
        assert_eq!(single_number(nums), expected);
    }
}
